ta/pkcs11/src/sanitize_object.c

```diff
@@ -0,0 +1,466 @@
+		len = sizeof(cli_ref) + cli_ref.size;
+
+		if (cli_ref.id == PKCS11_CKA_CLASS) {
+			uint32_t class;
+
+			if (cli_ref.size != sizeof(class)) {
+				rc = PKCS11_CKR_TEMPLATE_INCONSISTENT;
+				goto err;
+			}
+
+			TEE_MemMove(&class, cur + sizeof(cli_ref),
+				    cli_ref.size);
+
+			if (class_found != PKCS11_CKO_UNDEFINED_ID &&
+			    class_found != class) {
+				EMSG("Conflicting class value");
+				rc = PKCS11_CKR_TEMPLATE_INCONSISTENT;
+				goto err;
+			}
+
+			class_found = class;
+			continue;
+		}
+
+		/* The attribute is a type-in-class */
+		if (pkcs11_attr_is_type(cli_ref.id)) {
+			uint32_t type = 0;
+
+			if (cli_ref.size != pkcs11_attr_is_type(cli_ref.id)) {
+				rc = PKCS11_CKR_TEMPLATE_INCONSISTENT;
+				goto err;
+			}
+
+			TEE_MemMove(&type, cur + sizeof(cli_ref), cli_ref.size);
+
+			if (type_found != PKCS11_CKK_UNDEFINED_ID &&
+			    type_found != type) {
+				EMSG("Conflicting type-in-class value");
+				rc = PKCS11_CKR_TEMPLATE_INCONSISTENT;
+				goto err;
+			}
+
+			type_found = type;
+		}
+	}
+
+	/* Sanity */
+	if (cur != end) {
+		EMSG("Unexpected alignment issue");
+		rc = PKCS11_CKR_FUNCTION_FAILED;
+		goto err;
+	}
+
+	if (class_found != PKCS11_CKO_UNDEFINED_ID) {
+		rc = add_attribute(dst, PKCS11_CKA_CLASS,
+				   &class_found, sizeof(uint32_t));
+		if (rc)
+			goto err;
+	}
+
+	if (type_found != PKCS11_CKK_UNDEFINED_ID) {
+		rc = add_attribute(dst, PKCS11_CKA_KEY_TYPE,
+				   &type_found, sizeof(uint32_t));
+	}
+
+	return PKCS11_CKR_OK;
+
+err:
+	trace_attributes_from_api_head("bad-template", src, src_size);
+
+	return rc;
+}
+
+static enum pkcs11_rc sanitize_boolprop(struct obj_attrs **dst,
+					struct pkcs11_attribute_head *cli_ref,
+					char *cur, uint32_t *boolprop_base,
+					uint32_t *sanity)
+{
+	int shift = 0;
+	uint32_t mask = 0;
+	uint32_t value = 0;
+	uint32_t *boolprop_ptr = NULL;
+	uint32_t *sanity_ptr = NULL;
+
+	/* Get the boolean property shift position and value */
+	shift = pkcs11_attr2boolprop_shift(cli_ref->id);
+	if (shift < 0)
+		return PKCS11_RV_NOT_FOUND;
+
+	if (shift >= PKCS11_BOOLPROPS_MAX_COUNT)
+		return PKCS11_CKR_FUNCTION_FAILED;
+
+	mask = 1 << (shift % 32);
+	if ((*(uint8_t *)(cur + sizeof(*cli_ref))) == PKCS11_TRUE)
+		value = mask;
+	else
+		value = 0;
+
+	/* Locate the current config value for the boolean property */
+	boolprop_ptr = boolprop_base + (shift / 32);
+	sanity_ptr = sanity + (shift / 32);
+
+	/* Error if already set to a different boolean value */
+	if (*sanity_ptr & mask && value != (*boolprop_ptr & mask))
+		return PKCS11_CKR_TEMPLATE_INCONSISTENT;
+
+	if (value)
+		*boolprop_ptr |= mask;
+	else
+		*boolprop_ptr &= ~mask;
+
+	/* Store the attribute inside the serialized data */
+	if (!(*sanity_ptr & mask)) {
+		enum pkcs11_rc rc = PKCS11_CKR_OK;
+		uint8_t pkcs11_bool = !!value;
+
+		rc = add_attribute(dst, cli_ref->id, &pkcs11_bool,
+				   sizeof(uint8_t));
+		if (rc)
+			return rc;
+	}
+
+	*sanity_ptr |= mask;
+
+	return PKCS11_CKR_OK;
+}
```

Comment:
In general, this function would benefit from a few comments (precondition/postcondition).

Reply:
I'll rewrite it with help of the `bitstring.h` macros and add a few comments.